libutils/logging.c

```diff
@@ -281,6 +285,123 @@ void LogToSystemLog(const char *msg, LogLevel level)
 }
 #endif  /* !__MINGW32__ */
 
+#ifdef HAVE_SYSTEMD_SD_JOURNAL_H
+/**
+ * Count the number of key-value pairs provided in the variable-length
+ * argument list until (but not including) the "MESSAGE" key.
+ */
+static inline size_t CountExtraPairs(va_list va)
+{
+    va_list args;
+    va_copy(args, va);
+
+    size_t count = 0;
+    for (char *key = va_arg(args, char *); !StringEqual(key, "MESSAGE");
+         key = va_arg(args, char *), count++)
+    {
+        va_arg(args, char *);
+    }
+
+    va_end(args);
+    return count;
+}
+
+/**
+ * Add the extra key-value pairs to the iovec.
+ */
+static inline void iovecAddExtraPairs(va_list args,
+                                      struct iovec *const iov,
+                                      const size_t num_pairs)
+{
+    for (size_t i = 0; i < num_pairs; i++)
+    {
+        const char *const key = va_arg(args, char *);
+        const char *const value = va_arg(args, char *);
+
+        char *const pair = StringFormat("%s=%s", key, value);
+        assert(pair != NULL);
```

Comment:
I believe `StringFormat()` never returns `NULL` because it uses `xvasprintf()` that kills the process in such a case, but more `asserts()` never hurt. :)